File: algorithm/programmers/LV2/타겟 넘버.swift
//
//  타겟 넘버.swift
//  algorithm
//
//  Created by Kihun SONG on 2022/05/26.
//

import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
	var count = 0

	func dfs(_ depth: Int, _ sum: Int) {
		if depth == numbers.count {
			if sum == target {
				count += 1
			}
			return
		}

		dfs(depth + 1, sum + numbers[depth])
		dfs(depth + 1, sum - numbers[depth])
	}

	dfs(0, 0)

	return count
}

File: algorithm_java/src/main/java/programmers/lv2/타겟넘버.java
package programmers.lv2;

public class 타겟넘버 {
    int answer = 0;

    public int solution(int[] numbers, int target) {

        dfs(numbers, target, 0, 0);

        return answer;
    }

    public void dfs(int numbers[], int target, int index, int sum) {
        if (index == numbers.length) {
            if (sum == target) {
                answer += 1;
            }
            return;
        }

        dfs(numbers, target, index + 1, sum + numbers[index]);
        dfs(numbers, target, index + 1, sum - numbers[index]);

    }


}
